File: frontend/src/lib/api-client.ts
/**
 * Typed API client for backend communication
 */

const getApiBaseUrl = () => {
  const envUrl = process.env.NEXT_PUBLIC_API_URL;
  if (!envUrl) return 'http://localhost:3001/api';
  if (envUrl.endsWith('/api')) return envUrl;
  return `${envUrl}/api`;
};

const API_BASE_URL = getApiBaseUrl();

// ─── Response Types ───────────────────────────────────────────

export interface ApiResponse<T = unknown> {
  data?: T;
  error?: string;
  message?: string;
}

export interface PaginatedResponse<T> {
  data: T[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    totalPages: number;
  };
}

// ─── Auth Types ───────────────────────────────────────────────

export interface AuthTokens {
  accessToken: string;
  refreshToken: string;
}

export interface User {
  id: string;
  email: string;
  fullName?: string;
  country?: string;
  createdAt: string;
  updatedAt: string;
}

export interface AuthResult {
  user: User;
  tokens: AuthTokens;
}

// ─── Portfolio Types ──────────────────────────────────────────

export interface Portfolio {
  id: string;
  userId: string;
  name: string;
  baseCurrency: string;
  description?: string;
  isDefault: boolean;
  createdAt: string;
  updatedAt: string;
  // Calculated fields (optional as they may not be present on creation, but will be on fetch)
  totalValue?: number;
  totalUnrealizedPnl?: number;
  totalRealizedPnl?: number;
  snapshotTotalValue?: number;
}

export interface PositionSummary {
  assetId: string;
  symbol: string;
  name: string;
  assetType: string;
  quantity: number;
  avgPrice: number;
  currentPrice: number;
  marketValue: number;
  costBasis: number;
  unrealizedPnl: number;
  realizedPnl: number;
  pnlPercent: number;
  weight: number;
}

export interface AllocationSlice {
  assetType: string;
  value: number;
  percentage: number;
}

export interface PerformancePoint {
  date: string;
  value: number;
  pnl: number;
}

export interface PortfolioSummary {
  id: string;
  name: string;
  baseCurrency: string;
  totalValue: number;
  totalUnrealizedPnl: number;
  totalRealizedPnl: number;
  dailyChange: number;
  dailyChangePercent: number;
  positions: PositionSummary[];
  allocation: AllocationSlice[];
  performance: PerformancePoint[];
}

// ─── Transaction Types ────────────────────────────────────────

export type TransactionSide = 'BUY' | 'SELL' | 'DEPOSIT' | 'WITHDRAWAL' | 'TRANSFER_IN' | 'TRANSFER_OUT' | 'INCOME' | 'FEE';

export interface Transaction {
  id: string;
  portfolioId: string;
  assetId: string;
  side: TransactionSide;
  quantity: number;
  price: number;
  transactionCurrency: string;
  grossAmount: number;
  feeAmount: number;
  feeCurrency?: string;
  tradeTime: string;
  createdAt: string;
  note?: string;
  asset?: Asset;
  userAccountId?: string;
  type?: string;
  status?: string;
}

export interface CreateTransactionInput {
  assetId: string;
  side: TransactionSide;
  quantity: number;
  price: number;
  transactionCurrency: string;
  date: string;
  userAccountId?: string;
  feeAmount?: number;
  feeCurrency?: string;
  note?: string;
}

// ─── Asset Types ──────────────────────────────────────────────

export type AssetType = 'CRYPTO' | 'FOREX' | 'COMMODITY' | 'FIAT';

export interface Asset {
  id: string;
  symbol: string;
  name: string;
  assetType: AssetType;
  coingeckoId?: string;
  baseCurrency?: string;
  quoteCurrency?: string;
  isActive: boolean;
}

export interface CoinSearchResult {
  id: string;
  name: string;
  symbol: string;
  market_cap_rank: number | null;
  thumb: string;
  large: string;
}

export interface CreateAssetInput {
  symbol: string;
  name: string;
  coingeckoId: string;
  assetType?: AssetType;
}

// ─── Alert Types ──────────────────────────────────────────────

export type AlertType = 'PRICE_ABOVE' | 'PRICE_BELOW' | 'PERCENT_CHANGE' | 'PORTFOLIO_DRAWDOWN' | 'TARGET_PNL';

export interface Alert {
  id: string;
  userId: string;
  portfolioId?: string;
  assetId?: string;
  alertType: AlertType;
  targetPrice?: number;
  conditionValue: number;
  lookbackWindowMinutes?: number;
  isActive: boolean;
  lastTriggeredAt?: string;
  createdAt: string;
  asset?: Asset;
  portfolio?: Portfolio;
}

export interface CreateAlertInput {
  portfolioId?: string;
  assetId?: string;
  alertType: AlertType;
  targetPrice?: number;
  conditionValue: number;
  lookbackWindowMinutes?: number;
}

// ─── Exchange Types ───────────────────────────────────────────

export interface Exchange {
  id: string;
  name: string;
  code: string;
  type: string;
  websiteUrl?: string;
  logoUrl?: string;
  isActive: boolean;
}

export interface UserAccount {
  id: string;
  userId: string;
  exchangeId: string;
  name: string;
  accountType: string;
  baseCurrency: string;
  isDefault: boolean;
  exchange?: Exchange;
}

// ─── Settings Types ───────────────────────────────────────────

export interface UserSettings {
  id: string;
  userId: string;
  baseCurrency: string;
  timezone: string;
  locale: string;
  darkMode: boolean;
}

// ─── API Client ───────────────────────────────────────────────

class ApiClient {
  private isRefreshing = false;
  private refreshQueue: Array<{ resolve: (value: boolean) => void }> = [];

  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<ApiResponse<T>> {
    const url = `${API_BASE_URL}${endpoint}`;
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      ...(options.headers as Record<string, string>),
    };

    const token = this.getAccessToken();
    if (token) {
      headers['Authorization'] = `Bearer ${token}`;
    }

    try {
      const response = await fetch(url, {
        ...options,
        headers,
        credentials: 'include',
      });

      // Handle 401 with token refresh
      if (response.status === 401 && token) {
        const refreshed = await this.refreshToken();
        if (refreshed) {
          // Retry with new token
          headers['Authorization'] = `Bearer ${this.getAccessToken()}`;
          const retryResponse = await fetch(url, { ...options, headers, credentials: 'include' });
          const retryData = await retryResponse.json();
          if (!retryResponse.ok) {
            return { error: retryData.message || `HTTP ${retryResponse.status}` };
          }
          return { data: retryData };
        }
        return { error: 'Session expired. Please login again.' };
      }

      // Check content type before parsing
      const contentType = response.headers.get('content-type');
      let data;
      
      if (contentType && contentType.includes('application/json')) {
        data = await response.json();
      } else {
        // If not JSON (e.g. 404 HTML page), treat as error text
        const text = await response.text();
        return {
          error: `API Error: ${response.status} ${response.statusText}`,
          message: text.substring(0, 100) // truncate potentially long HTML
        };
      }

      if (!response.ok) {
        return {
          error: data.message || `HTTP ${response.status}: ${response.statusText}`,
          message: data.message,
        };
      }

      return { data };
    } catch (error) {
      console.error('API request failed:', error);
      return {
        error: error instanceof Error ? error.message : 'Network error',
      };
    }
  }

  // ─── Auth Methods ───────────────────────────────────────────

  async register(email: string, password: string, fullName?: string) {
    return this.request<AuthResult>('/auth/register', {
      method: 'POST',
      body: JSON.stringify({ email, password, fullName }),
    });
  }

  async login(email: string, password: string) {
    return this.request<AuthResult>('/auth/login', {
      method: 'POST',
      body: JSON.stringify({ email, password }),
    });
  }

  async refreshToken(): Promise<boolean> {
    if (this.isRefreshing) {
      return new Promise<boolean>((resolve) => {
        this.refreshQueue.push({ resolve });
      });
    }

    this.isRefreshing = true;
    const refreshTokenValue = this.getRefreshToken();

    if (!refreshTokenValue) {
      this.isRefreshing = false;
      return false;
    }

    try {
      const response = await fetch(`${API_BASE_URL}/auth/refresh`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ refreshToken: refreshTokenValue }),
      });

      if (!response.ok) {
        this.clearTokens();
        this.resolveRefreshQueue(false);
        return false;
      }

      const tokens: AuthTokens = await response.json();
      this.setTokens(tokens);
      this.resolveRefreshQueue(true);
      return true;
    } catch {
      this.clearTokens();
      this.resolveRefreshQueue(false);
      return false;
    } finally {
      this.isRefreshing = false;
    }
  }

  private resolveRefreshQueue(success: boolean) {
    this.refreshQueue.forEach(({ resolve }) => resolve(success));
    this.refreshQueue = [];
  }

  async logout() {
    try {
      await this.request('/auth/logout', { method: 'POST' });
    } catch {
      // Ignore logout errors
    }
  }

  async getCurrentUser() {
    return this.request<User>('/users/me');
  }

  // ─── Portfolio Methods ──────────────────────────────────────

  async getPortfolios() {
    return this.request<Portfolio[]>('/portfolios');
  }

  async createPortfolio(name: string, baseCurrency = 'USD') {
    return this.request<Portfolio>('/portfolios', {
      method: 'POST',
      body: JSON.stringify({ name, baseCurrency }),
    });
  }

  async getPortfolio(id: string) {
    return this.request<Portfolio>(`/portfolios/${id}`);
  }

  async updatePortfolio(id: string, data: { name?: string; baseCurrency?: string; description?: string }) {
    return this.request<Portfolio>(`/portfolios/${id}`, {
      method: 'PUT',
      body: JSON.stringify(data),
    });
  }

  async deletePortfolio(id: string) {
    return this.request<void>(`/portfolios/${id}`, { method: 'DELETE' });
  }

  async getPortfolioSummary(id: string) {
    return this.request<PortfolioSummary>(`/portfolios/${id}/summary`);
  }

  async getPortfolioPerformance(id: string, days = 30) {
    return this.request<PerformancePoint[]>(`/portfolios/${id}/performance?days=${days}`);
  }

  async getTopPerformers(id: string, limit = 5) {
    return this.request<PositionSummary[]>(`/portfolios/${id}/top-performers?limit=${limit}`);
  }

  // ─── Transaction Methods ────────────────────────────────────

  async createTransaction(portfolioId: string, transaction: CreateTransactionInput) {
    return this.request<Transaction>(`/portfolios/${portfolioId}/transactions`, {
      method: 'POST',
      body: JSON.stringify(transaction),
    });
  }

  async getPortfolioTransactions(portfolioId: string, page = 1, limit = 50) {
    return this.request<PaginatedResponse<Transaction>>(`/portfolios/${portfolioId}/transactions?page=${page}&limit=${limit}`);
  }

  async deleteTransaction(portfolioId: string, transactionId: string) {
    return this.request<void>(`/portfolios/${portfolioId}/transactions/${transactionId}`, {
      method: 'DELETE',
    });
  }

  // ─── Asset Methods ──────────────────────────────────────────

  async getAssets() {
    return this.request<Asset[]>('/assets');
  }

  async createAsset(asset: CreateAssetInput) {
    return this.request<Asset>('/assets', {
      method: 'POST',
      body: JSON.stringify(asset),
    });
  }

  async searchAssets(query: string) {
    return this.request<CoinSearchResult[]>(`/assets/search?query=${encodeURIComponent(query)}`);
  }

  // ─── Alert Methods ──────────────────────────────────────────

  async getAlerts() {
    return this.request<Alert[]>('/alerts');
  }

  async createAlert(alert: CreateAlertInput) {
    return this.request<Alert>('/alerts', {
      method: 'POST',
      body: JSON.stringify(alert),
    });
  }

  async updateAlert(id: string, data: Partial<CreateAlertInput> & { isActive?: boolean }) {
    return this.request<Alert>(`/alerts/${id}`, {
      method: 'PUT',
      body: JSON.stringify(data),
    });
  }

  async deleteAlert(id: string) {
    return this.request<void>(`/alerts/${id}`, { method: 'DELETE' });
  }

  // ─── Exchange Methods ───────────────────────────────────────

  async getExchanges() {
    return this.request<Exchange[]>('/exchanges');
  }

  async getUserAccounts() {
    return this.request<UserAccount[]>('/exchanges/accounts');
  }

  // ─── Settings Methods ───────────────────────────────────────

  async getSettings() {
    return this.request<UserSettings>('/users/settings');
  }

  async updateSettings(data: Partial<UserSettings>) {
    return this.request<UserSettings>('/users/settings', {
      method: 'PUT',
      body: JSON.stringify(data),
    });
  }

  // ─── Token Management ──────────────────────────────────────

  getAccessToken(): string | null {
    if (typeof window === 'undefined') return null;
    return localStorage.getItem('access_token');
  }

  private getRefreshToken(): string | null {
    if (typeof window === 'undefined') return null;
    return localStorage.getItem('refresh_token');
  }

  setTokens(tokens: AuthTokens) {
    if (typeof window === 'undefined') return;
    localStorage.setItem('access_token', tokens.accessToken);
    localStorage.setItem('refresh_token', tokens.refreshToken);
  }

  clearTokens() {
    if (typeof window === 'undefined') return;
    localStorage.removeItem('access_token');
    localStorage.removeItem('refresh_token');
  }

  isAuthenticated(): boolean {
    return !!this.getAccessToken();
  }
}

export const apiClient = new ApiClient();

File: app/lib/core/api/api_client.dart
import 'package:dio/dio.dart';
import '../storage/secure_storage.dart';
import '../models/user.dart';

class ApiClient {
  static final ApiClient _instance = ApiClient._internal();
  factory ApiClient() => _instance;

  late final Dio _dio;
  final SecureStorage _storage = SecureStorage();
  String? _token;

  // Default to localhost, can be overridden via env
  static const String _defaultBaseUrl = 'http://localhost:3001/api';

  ApiClient._internal() {
    _dio = Dio(
      BaseOptions(
        baseUrl: _defaultBaseUrl,
        connectTimeout: const Duration(seconds: 10),
        receiveTimeout: const Duration(seconds: 10),
        headers: {'Content-Type': 'application/json'},
      ),
    );

    _dio.interceptors.add(InterceptorsWrapper(
      onRequest: (options, handler) async {
        _token ??= await _storage.getAccessToken();
        if (_token != null) {
          options.headers['Authorization'] = 'Bearer $_token';
        }
        handler.next(options);
      },
      onError: (DioException error, handler) async {
        if (error.response?.statusCode == 401 &&
            _token != null &&
            !error.requestOptions.path.contains('/auth/login') &&
            !error.requestOptions.path.contains('/auth/refresh')) {
          try {
            final refreshToken = await _storage.getRefreshToken();
            if (refreshToken == null) {
              await logout();
              return handler.next(error);
            }

            // Attempt to refresh token
            final response = await _dio.post(
              '/auth/refresh',
              data: {'refreshToken': refreshToken},
            );

            if (response.statusCode == 200 && response.data != null) {
              final newAccessToken = response.data['accessToken'];
              final newRefreshToken = response.data['refreshToken'];

              if (newAccessToken != null && newRefreshToken != null) {
                await _storage.saveTokens(
                  accessToken: newAccessToken,
                  refreshToken: newRefreshToken,
                );
                _token = newAccessToken;

                // Update the failed request's header with new token
                error.requestOptions.headers['Authorization'] =
                    'Bearer $newAccessToken';

                // Retry the request
                final cloneReq = await _dio.fetch(error.requestOptions);
                return handler.resolve(cloneReq);
              }
            }
          } catch (e) {
            // Refresh failed, user must login again
            await logout();
            return handler.next(error);
          }
        }
        handler.next(error);
      },
    ));
  }

  void setBaseUrl(String url) {
    _dio.options.baseUrl = url;
  }

  // ─── Auth ──────────────────────────────────────────────────────────

  Future<Map<String, dynamic>> register(String email, String password,
      {String? fullName}) async {
    try {
      final response = await _dio.post('/auth/register', data: {
        'email': email,
        'password': password,
        if (fullName != null) 'fullName': fullName,
      });
      final data = response.data;
      if (data['tokens'] != null) {
        final tokens = AuthTokens.fromJson(data['tokens']);
        await _storage.saveTokens(
            accessToken: tokens.accessToken, refreshToken: tokens.refreshToken);
        _token = tokens.accessToken;
      }
      return {'data': data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> login(String email, String password) async {
    try {
      final response = await _dio.post('/auth/login', data: {
        'email': email,
        'password': password,
      });
      final data = response.data;
      if (data['tokens'] != null) {
        final tokens = AuthTokens.fromJson(data['tokens']);
        await _storage.saveTokens(
            accessToken: tokens.accessToken, refreshToken: tokens.refreshToken);
        _token = tokens.accessToken;
      }
      return {'data': data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<void> logout() async {
    await _storage.clearTokens();
    _token = null;
  }

  Future<Map<String, dynamic>> getCurrentUser() async {
    try {
      final response = await _dio.get('/users/me');
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  // ─── Portfolios ────────────────────────────────────────────────────

  Future<Map<String, dynamic>> getPortfolios() async {
    try {
      final response = await _dio.get('/portfolios');
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> createPortfolio(String name,
      {String baseCurrency = 'USD'}) async {
    try {
      final response = await _dio.post('/portfolios', data: {
        'name': name,
        'baseCurrency': baseCurrency,
      });
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> getPortfolio(String id) async {
    try {
      final response = await _dio.get('/portfolios/$id');
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> getPortfolioSummary(String id) async {
    try {
      final response = await _dio.get('/portfolios/$id/summary');
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> getPortfolioTransactions(String id) async {
    try {
      final response = await _dio.get('/portfolios/$id/transactions');
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> createTransaction(
      String portfolioId, Map<String, dynamic> data) async {
    try {
      final response =
          await _dio.post('/portfolios/$portfolioId/transactions', data: data);
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> updateTransaction(String transactionId,
      String portfolioId, Map<String, dynamic> data) async {
    try {
      final response = await _dio.put(
          '/portfolios/$portfolioId/transactions/$transactionId',
          data: data);
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> deleteTransaction(
      String transactionId, String portfolioId) async {
    try {
      await _dio.delete('/portfolios/$portfolioId/transactions/$transactionId');
      return {'data': null};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> updatePortfolio(String id, String name,
      {String? baseCurrency}) async {
    try {
      final response = await _dio.put('/portfolios/$id', data: {
        'name': name,
        if (baseCurrency != null) 'baseCurrency': baseCurrency,
      });
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> deletePortfolio(String id) async {
    try {
      await _dio.delete('/portfolios/$id');
      return {'data': null};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  // ─── Alerts ────────────────────────────────────────────────────────

  Future<Map<String, dynamic>> getAlerts() async {
    try {
      final response = await _dio.get('/alerts');
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> createAlert(Map<String, dynamic> data) async {
    try {
      final response = await _dio.post('/alerts', data: data);
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> updateAlert(
      String id, Map<String, dynamic> data) async {
    try {
      final response = await _dio.put('/alerts/$id', data: data);
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> deleteAlert(String id) async {
    try {
      await _dio.delete('/alerts/$id');
      return {'data': null};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  // ─── Assets ────────────────────────────────────────────────────────

  Future<Map<String, dynamic>> getAssets({String? query}) async {
    try {
      final response = await _dio.get('/assets', queryParameters: {
        if (query != null && query.isNotEmpty) 'search': query,
      });
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  // ─── Market Data ───────────────────────────────────────────────────

  Future<Map<String, dynamic>> getMarketData({
    List<String>? ids,
    String currency = 'usd',
    bool sparkline = false,
  }) async {
    try {
      final response = await _dio.get('/prices/market-data', queryParameters: {
        if (ids != null && ids.isNotEmpty) 'ids': ids.join(','),
        'currency': currency,
        'sparkline': sparkline,
      });
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> searchCoins(String query) async {
    try {
      final response = await _dio.get('/prices/search', queryParameters: {
        'q': query,
      });
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  Future<Map<String, dynamic>> getPriceHistory(String assetId,
      {String interval = '1d'}) async {
    try {
      final response =
          await _dio.get('/prices/$assetId/history', queryParameters: {
        'interval': interval,
      });
      return {'data': response.data};
    } on DioException catch (e) {
      return {'error': e.response?.data?['message'] ?? e.message};
    }
  }

  bool isAuthenticated() => _token != null;
}

final apiClient = ApiClient();
